lib/VM.php

```diff
@@ -279,6 +279,11 @@ public function run(Block $block): int {
                     $dst = $frame->scope[$op->arg1];
                     $dst->bool($value);
                     break;
+                case OpCode::TYPE_INCLUDE:
+                    $file = $frame->scope[$op->arg1]->toString();
+                    $parsed = $this->context->runtime->parseAndCompileFile($file);
+		    $this->context->runtime->run($parsed);
```

Comment:
Running it doesn't need to come out of the VM. You ideally should be able to just jump to the new context, pushing the old frame (so returns go back to the right place).

Not tested, but something like this:

```suggestion
		    $frame = $parsed->getFrame($this->context);
		    $this->context->push($frame);
                    $frame = $new;
                    goto restart;
```

Reply:
I've been trying to get that to work, but context popping seems to be corrupting the parent frame and I can't figure out where.

After an include, when it hits the implicit "TYPE_RETURN_VOID" opcode that the CFG pass adds, the implicit return goes to the next frame. The next frame gets popped (the is_null check doesn't get executed), but $frame->block->nOpCodes and $frame->pos remain the values from the included block. As a result, the while loop condition fails and the original frame isn't picking up where it left off, but exiting after the included file.

If I dump $frame->block after the while loop, it does seem to be the block of the included file, not the original, but I'm not sure how

Reply:
nevermind, I was doing something stupid and pushing the wrong frame. Now the only problem is that side-effects from the included frame.